### ISSUE TITLE
nightly builds are timing out

### ISSUE THREAD

**Issue by epaulson** [BrickSchema/Brick#201]
The nightly build started timing out about 10 days ago.

Pip released a new version about 10 days ago with a new resolver:
http://pyfound.blogspot.com/2020/11/pip-20-3-new-resolver.html

Perhaps related?

**Comment by epaulson** [comment]
The change of ubuntu-latest to 20.04 started on the very same day:
https://github.com/actions/virtual-environments/issues/1816

but it doesn't look that runner change has rolled out to Brick yet - the logs look like the past few days are still running on 18.04

**Comment by gtfierro** [comment]
What's interesting to me is that it only seems to fail for Python 3.8.5; this is true in your fix https://github.com/BrickSchema/Brick/pull/200 as well as the [actual build logs](https://github.com/BrickSchema/Brick/runs/1534730885?check_suite_focus=true). It looks to be something related to Pip at first glance, but definitely worth looking into

**Comment by gtfierro** [comment]
One thought -- the versions of dependencies in our `requirements.txt` might have trouble resolving (looks to be a similar issue [here](https://github.com/pantsbuild/pants/issues/10885))

**Comment by epaulson** [comment]
I can reproduce it on my 18.04 VM with Python 3.8.6 installed from apt 
(using this https://askubuntu.com/questions/1197683/how-do-i-install-python-3-8-in-lubuntu-18-04 to get a working venv)

The github setup-python action does as a last step in its installer script:
```

**Comment by epaulson** [comment]
I can reproduce it on my 18.04 VM with Python 3.8.6 installed from apt 
(using this https://askubuntu.com/questions/1197683/how-do-i-install-python-3-8-in-lubuntu-18-04 to get a working venv)

The github setup-python action does as a last step in its installer script:
```
./python -m ensurepip
./python -m pip install --ignore-installed pip
```
so each build always picks up the latest and greatest pip (so much for reproducible build environments)

The new pip doesn't like our requirements.txt - you can see in the log files that since the 30th it's downloading every released version of pre-commit, flake8, docker, and brickschema among others, and that's slowing down as more versions are downloaded - not sure if pypi is throttling us or if the resolver in pip has some algorithmic limitation like its n^2 in the number of versions it's considering or something. 

Maybe we should just freeze the versions we're using in requirements.txt? In my brief glance through the pip documentation it still seemed like our approach should still work, but maybe one of the packages is giving it trouble.

**Comment by gtfierro** [comment]
Good find! I think that's likely the issue. I'm seeing the following in the logs

```
INFO: pip is looking at multiple versions of flake8 to determine which version is compatible with other requirements.
```
and

```
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime.
```

Stricter/frozen requirements are likely the fix. I can try to take some time this afternoon to figure out some frozen versions that fix this, but let me know if you get to it first

**Comment by epaulson** [comment]
There's a large thread in pip re: this
https://github.com/pypa/pip/issues/9187

( 9215 similar )

The documentation to read is here: 
https://pip.pypa.io/en/latest/user_guide/#dependency-resolution-backtracking

**Comment by epaulson** [comment]
tl;dr: we're explicitly asking a git checkout that gives us OWL-RL 5.2.0 but pyshacl wants OWL-RL 5.2.1
```
Collecting git+https://github.com/RDFLib/OWL-RL.git@471d1dfe8f6 (from -r requirements.txt (line 3))
  Cloning https://github.com/RDFLib/OWL-RL.git (to revision 471d1dfe8f6) to /tmp/pip-req-build-estr88w0
  WARNING: Did not find branch or tag '471d1dfe8f6', assuming revision or ref.
```

I just hardcoded a couple of versions so pip would stop backtracking so we could get a better error message. This is what pip says with my first try:

```
ERROR: Cannot install -r requirements.txt (line 10), -r requirements.txt (line 3) and rdflib-jsonld==0.4.0 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested rdflib-jsonld==0.4.0
    owlrl 5.2.0 depends on rdflib-jsonld>=0.4.0
    pyshacl 0.12.2 depends on rdflib-jsonld<0.6.0 and >=0.5.0
    The user requested rdflib-jsonld==0.4.0
    owlrl 5.2.0 depends on rdflib-jsonld>=0.4.0
    pyshacl 0.12.1.post2 depends on rdflib-jsonld<0.6.0 and >=0.5.0
    The user requested rdflib-jsonld==0.4.0
    owlrl 5.2.0 depends on rdflib-jsonld>=0.4.0
    pyshacl 0.12.1.post1 depends on rdflib-jsonld<0.6.0 and >=0.5.0
    The user requested rdflib-jsonld==0.4.0
    owlrl 5.2.0 depends on rdflib-jsonld>=0.4.0
    pyshacl 0.12.1 depends on rdflib-jsonld<0.6.0 and >=0.5.0

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict
```

This is the requirements.txt I'm using, I don't think any of which matter for the conflicts we're hitting:
```
isodate==0.6.0
owlrl>=5.2.0
git+https://github.com/RDFLib/OWL-RL.git@471d1dfe8f6 # because of issue https://github.com/RDFLib/OWL-RL/issues/29, use git directly until it's released.
pyparsing==2.3.1
rdflib>=5.0.0
rdflib-jsonld==0.4.0
six==1.12.0
pytest>=5.4.3
tqdm>=4.46.1
pyshacl>=0.12.0
docker>=4.3.0
brickschema>=0.1.8
black==19.10b0
pre-commit==2.4.0
flake8==3.8.2
semver>=2.10.1
```
(I fixed versions for pytest, tqdm, black, flake8, pre-commit, and semver)

I set the rdflib-jsonld version to 0.5.0 to make some progress, and now we're hung up in OWL-RL:

```
ERROR: Cannot install -r requirements.txt (line 10), -r requirements.txt (line 12), owlrl 5.2.0 (from git+https://github.com/RDFLib/OWL-RL.git@471d1dfe8f6) and owlrl>=5.2.0 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested owlrl>=5.2.0
    The user requested owlrl 5.2.0 (from git+https://github.com/RDFLib/OWL-RL.git@471d1dfe8f6)
    brickschema 0.1.8 depends on owlrl<6.0 and >=5.2
    pyshacl 0.12.2 depends on owlrl<6.0.0 and >=5.2.1
    The user requested owlrl>=5.2.0
    The user requested owlrl 5.2.0 (from git+https://github.com/RDFLib/OWL-RL.git@471d1dfe8f6)
    brickschema 0.1.8 depends on owlrl<6.0 and >=5.2
    pyshacl 0.12.1.post2 depends on owlrl<6.0.0 and >=5.2.1
    The user requested owlrl>=5.2.0
    The user requested owlrl 5.2.0 (from git+https://github.com/RDFLib/OWL-RL.git@471d1dfe8f6)
    brickschema 0.1.8 depends on owlrl<6.0 and >=5.2
    pyshacl 0.12.1.post1 depends on owlrl<6.0.0 and >=5.2.1
    The user requested owlrl>=5.2.0
    The user requested owlrl 5.2.0 (from git+https://github.com/RDFLib/OWL-RL.git@471d1dfe8f6)
    brickschema 0.1.8 depends on owlrl<6.0 and >=5.2
    pyshacl 0.12.1 depends on owlrl<6.0.0 and >=5.2.1

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

```

Is there anything special about that particular checkin in OWL-RL, or did we just need a pre-release and we can relax that requirement?

(We should still probably spec some minimums for the other packages just to keep pip from downloading a slug of packages)

**Comment by gtfierro** [comment]
I think there was a bug or performance regression in that version? Or it might have been a holdover from some development --- in any case, we should definitely remove that restriction and accept `owlrl 6.0.0` at least

Nice work, and thanks for digging into this!

**Comment by epaulson** [comment]
In fact, pip has been warning us about this for a while - in nighty build #101 on November 29th

```
We recommend you use --use-feature=2020-resolver to test your packages with the new resolver before it becomes the default.

pyshacl 0.14.0 requires rdflib-jsonld<0.6.0,>=0.5.0, but you'll have rdflib-jsonld 0.4.0 which is incompatible.
brickschema 0.1.10 requires pyshacl<0.13.0,>=0.12.1, but you'll have pyshacl 0.14.0 which is incompatible.
brickschema 0.1.10 requires pytest<6.0,>=5.3, but you'll have pytest 6.1.2 which is incompatible.
```
